### ISSUE TITLE
[custom serde annotations] the trait `Deserialize<'_>` is not implemented for `MessageField<T>`

### ISSUE THREAD

**Issue by svanharmelen** [stepancheg/rust-protobuf#703]
I also just opened #702 as well as my goal is to parse both protobuf and JSON payloads using the same generated structs using either `protobuf-json-mapping` or custom Serde annotations. The other issue describes an issue with `protobuf-json-mapping` and this issue is used to describe an issue with using custom Serde annotations.

To make it easy to reproduce the issue, I created a debug-example repo with a "working" example giving the error: https://github.com/svanharmelen/debug-example/tree/serde (notice this example is in the `serde` branch)

I'm using an update proto file:

```proto
syntax = "proto2";

message TestMessage {
  message SubMessage {
    optional string name = 1;
  }

  message MainMessage {
    optional string name           = 1;
    optional SubMessage submessage = 2;

    oneof value {
      uint32   int_value    = 3;
      uint64   long_value   = 4;
      float    float_value  = 5;
      double   double_value = 6;
    }
  }
}

```

But this time I am using this `lib.rs`:

```rust
include!(concat!(env!("OUT_DIR"), "/protos/mod.rs"));

#[cfg(test)]
mod test {
    use crate::example::test_message::{main_message::Value, MainMessage};

    #[test]
    fn test() {
        let json = r#"{
          "name": "test",
          "value": 10.3
        }"#;

        let message: MainMessage = serde_json::from_str(json).expect("failed to parse JSON");

        assert_eq!(message.name, Some("test".to_string()));
        assert_eq!(message.value, Some(Value::FloatValue(10.3)));
    }
}
```

The error I get is about `MessageField<T>` types not implementing the `Deserialize` trait and I have now idea how I should fix that in a generic way (the actual proto file I use has a bunch of these embedded submessages):

```
3897 |         pub properties: ::protobuf::MessageField<PropertySet>,
     |         ^^^ the trait `Deserialize<'_>` is not implemented for `MessageField<PropertySet>`
```

I'm fine to use either Serde annotations or `protobuf-json-mapping`, whichever one is easier to fix will do I guess 😉

Thanks!

**Comment by svanharmelen** [comment]
OK, so I just managed to find a solution for the main issue described above. Updating this (in `build.rs`:

```rust
        fn oneof(&self, _oneof: &OneofDescriptor) -> Customize {
            Customize::default().before("#[derive(::serde::Deserialize)]")
        }
```

To this:

```rust
        fn oneof(&self, _oneof: &OneofDescriptor) -> Customize {
            Customize::default().before("#[derive(::serde::Deserialize)]\n#[serde(untagged)]")
        }
```

Seems to solve the problem! So now the only issue I have left stopping me to use custom Serde annotations are the `MessageField<T>` errors:

```
3897 |         pub properties: ::protobuf::MessageField<PropertySet>,
         |         ^^^ the trait `Deserialize<'_>` is not implemented for `MessageField<PropertySet>`
```

It would be really awesome if you (or anyone else reading this message) can point me in the right direction on how to sove these issues...

**Comment by svanharmelen** [comment]
I just updated the example in https://github.com/svanharmelen/debug-example/tree/serde to include the fix I just found and to demonstrate the `MessageField<T>` related errors.

**Comment by stepancheg** [comment]
I think some serde annotation can be inserted to specify message field should be parsed/serialized with custom function.

**Comment by svanharmelen** [comment]
Yes, I was thinking the same... But I cannot find a way to "detect" which field are of type `MessageField<T>` in the available `CustomizeCallback` methods in order to write an annotation on them. Any suggestions?

**Comment by icdevin** [comment]
@svanharmelen I'm not sure if you ever figured this out, but I have some solution, although I am using it to skip adding derives on MessageField<T>.

The `fn message` bit adds the SerDe derive above the outer message type, then I skip on inner MessageField, and special fields.

```
struct GenSerde;
impl CustomizeCallback for GenSerde {
  fn message(&self, _message: &MessageDescriptor) -> Customize {
    Customize::default().before("#[derive(::serde::Serialize, ::serde::Deserialize)]")
  }

  // Skips adding the derive macro above field-level Message types, as they
  // are wrapped in a MessageField struct which is not Serialize or Deserialize
  fn field(&self, field: &FieldDescriptor) -> Customize {
    if field.proto().type_() == Type::TYPE_MESSAGE {
      Customize::default().before("#[serde(skip)]")
    } else {
      Customize::default()
    }
  }

  fn special_field(&self, _message: &MessageDescriptor, _field: &str) -> Customize {
    Customize::default().before("#[serde(skip)]")
  }
}
```

**Comment by svanharmelen** [comment]
Thanks for your reply @icdevin, much appreciated!

I ended up making it very specific by using the field name. Feels a bit nasty, but it does solve the issue 😏 If I remember correctly I did try your suggestion, but that ended up skipping more field then needed. Might check it again when I have to update or make changes to the crate again...

**Comment by impy88** [comment]
Hi @svanharmelen,

I solved the issue using [remote derive in serde](https://serde.rs/remote-derive.html)

In generator code you can use that snippet to annotate every `MessageField` with serde derive. `is_singular()` is really important here as you don't want to cover `Vec`. 

```rust

fn field(&self, field: &FieldDescriptor) -> Customize {
    if field.proto().type_() == Type::TYPE_MESSAGE && field.is_singular() {
        Customize::default()
             .before("#[serde(with = \"crate::MessageFieldDef\")]")
   } 
}
```

and in your crate you'd need to implement `MessageFieldDef` in a following way

```rust
#[derive(Serialize, Deserialize)]
#[serde(remote = "MessageField")]
pub struct MessageFieldDef<T>(pub Option<Box<T>>);
```

**Comment by svanharmelen** [comment]
Thanks for the tip @impy88! When time permits I'll have a look at this approach to see if/how it works in our setup 👍🏻

**Comment by svanharmelen** [comment]
@impy88 you're a life-safer!

I just came to a point where I needed to update some things and could no longer just skip the field like I did before (as I now actually want to use the field) and your suggestion helped me out enormously!

I did need to make a few more tweaks to get it working for our exact use case, but got everything working within an hour or fiddling and testing so pretty happy with that.

So big "thank you" for taking the time and effort to reply to my (already closed) issue ❤️
